src/alda/core.clj

```diff
@@ -1024,3 +1025,167 @@
+
+(defmethod event-map->record {:type "attribute-update"
+                              :attribute "volume"}
+  [{:keys [value]}]
+  (volume value))
+
+(defmethod event-map->record {:type "attribute-update"
+                              :attribute "track-volume"}
+  [{:keys [value]}]
+  (track-volume value))
+
+(defmethod event-map->record {:type "attribute-update"
+                              :attribute "panning"}
+  [{:keys [value]}]
+  (panning value))
+
+(defmethod event-map->record {:type "attribute-update"
+                              :attribute "quantization"}
+  [{:keys [value]}]
+  (quantization value))
+
+(defmethod event-map->record {:type "attribute-update"
+                              :attribute "key-signature"}
+  [{:keys [value]}]
+  (key-signature value))
+
+(defmethod event-map->record {:type "attribute-update"
+                              :attribute "transposition"}
+  [{:keys [value]}]
+  (transposition value))
+
+(defmethod event-map->record {:type "attribute-update"
+                              :attribute "reference-pitch"}
+  [{:keys [value]}]
+  (reference-pitch value))
+
+(defmethod event-map->record {:type "global-attribute-update"
+                              :attribute "octave"}
+  [{:keys [value]}]
+  (octave! value))
+
+(defmethod event-map->record {:type "global-attribute-update"
+                              :attribute "tempo"}
+  [{:keys [value]}]
+  (tempo! value))
+
+(defmethod event-map->record {:type "global-attribute-update"
+                              :attribute "volume"}
+  [{:keys [value]}]
+  (volume! value))
+
+(defmethod event-map->record {:type "global-attribute-update"
+                              :attribute "track-volume"}
+  [{:keys [value]}]
+  (track-volume! value))
+
+(defmethod event-map->record {:type "global-attribute-update"
+                              :attribute "panning"}
+  [{:keys [value]}]
+  (panning! value))
+
+(defmethod event-map->record {:type "global-attribute-update"
+                              :attribute "quantization"}
+  [{:keys [value]}]
+  (quantization! value))
+
+(defmethod event-map->record {:type "global-attribute-update"
+                              :attribute "key-signature"}
+  [{:keys [value]}]
+  (key-signature! value))
+
+(defmethod event-map->record {:type "global-attribute-update"
+                              :attribute "transposition"}
+  [{:keys [value]}]
+  (transposition! value))
+
+(defmethod event-map->record {:type "global-attribute-update"
+                              :attribute "reference-pitch"}
+  [{:keys [value]}]
+  (reference-pitch! value))
+
+(defn parse-events!
+  "Converts its arguments into a string of Alda code (via [[->str]]) and sends
+   it to the Alda CLI to be parsed into events JSON.
+   
+   Returns a seq of deserialized records."
+  [& xs]
+  (map event-map->record
+       (json/read-value
+        (alda "parse" "--output" "events" "--code" (->str xs))
+        json/keyword-keys-object-mapper)))
```

Comment:
I think this function might be doing a little bit too much, in that it accepts a seq of Alda event records. I think it might be better for this function to take a string as an argument instead. That way, we have the flexibility to call it with either a string:

```clojure
(parse-events "piano: c d e")
```

Or we could call it with events, if desired:

```clojure
(parse-events (->str (chord (note ...) (note ...)))
```

I'd also suggest calling it `parse-events` (without the `!`) because what we care about is the return value, not any particular side effects.

Reply:
Ah I modeled it off of `play!`, hence the `[& xs]` argument interface and `!` convention. Given that `->str` also works for strings, I think the current implementation is the most generic -- it should work for `(parse-events (->str (chord (note ...) (note ...)))` or `(parse-events (chord (note ...) (note ...))` just like `play!`. Happy to change this to only accept a string if you want though. :smiley:

Reply:
Alright, you've sold me on this. I like `parse-events` is consistent with `play!` in that both functions have the same signature, and both functions are doing essentially the same thing:

* stringify their arguments
* hand them as input to an `alda` CLI command

And like you mentioned, `->str` is a no-op for strings, so it's not like making this function explicitly accept a string would provide any additional flexibility.